hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/VolumeSet.java

```diff
@@ -239,23 +249,24 @@ private void checkAllVolumes() throws IOException {
    * @param failedVolumes
    */
   private void handleVolumeFailures(Set<HddsVolume> failedVolumes) {
-    for (HddsVolume v: failedVolumes) {
-      this.writeLock();
-      try {
+    this.writeLock();
+    try {
+      for (HddsVolume v : failedVolumes) {
         // Immediately mark the volume as failed so it is unavailable
         // for new containers.
-        volumeMap.remove(v.getHddsRootDir().getPath());
-        failedVolumeMap.putIfAbsent(v.getHddsRootDir().getPath(), v);
-      } finally {
-        this.writeUnlock();
+        failVolume(v.getHddsRootDir().getPath());
       }
-
-      // TODO:
-      // 1. Mark all closed containers on the volume as unhealthy.
-      // 2. Consider stopping IO on open containers and tearing down
-      //    active pipelines.
-      // 3. Handle Ratis log disk failure.
+      if (failedVolumeListener != null) {
```

Comment:
This can probably be done outside the writeLock - should be safe.
also, we can invoke the listener outside the for loop - perhaps like:
if (!failedVolumes.isEmpty()) {failedVolumeListener.run();}

Reply:
Yes we can move it outside. But I think there is a race condition in marking a volume as failed. A container in process of creation can still get created on the failed volume. We can handle it as part of separate jira though.